File: HW4/hw4.cpp
#include <iostream>
#include <string>
#include <stack>
#include <bits/stdc++.h>
using namespace std;    
        
int longestRun(vector<bool> boolVec) {
    int longestAnswer = 0;
    int answer = 0;

    for (int i = 0; i < boolVec.size(); i++) {
        if (boolVec[i]) {
            answer++;
        } else {
            longestAnswer = max(longestAnswer, answer);
            answer = 0;
        }
    }

    return max(longestAnswer, answer);
}

class Tree {
    public:
        unsigned value;
        vector<Tree *> children;
        Tree(unsigned value, vector<Tree *> children) {
        this->value = value;
        this->children = children;
    }
};


unsigned maxTreeValue(Tree* root) {
    if (root == nullptr) {
        return 0;
    }

    unsigned maxValue = root->value;
    stack<Tree*> nodes;
    nodes.push(root);

    while (!nodes.empty()) {
        Tree* current = nodes.top();
        nodes.pop();

        // Update maxValue if the current node's value is greater
        maxValue = max(maxValue, current->value);

        // Add all children to the stack
        for (Tree* child : current->children) {
            if (child != nullptr) {
                nodes.push(child);
            }
        }
    }

    return maxValue;
}

int main() {
    vector<bool> boolVec = {true, true, false, true, true, true, false};
    cout << longestRun(boolVec) << endl;
}

File: HW4/hw4.hs
import Data.List (group)

longest_run :: [Bool] -> Int
longest_run xs = maximum (0 : map length (filter (all (== True)) (group xs)))

data Tree = Empty | Node Integer [Tree]

max_tree_value :: Tree -> Integer
max_tree_value Empty = 0
max_tree_value (Node value subtrees) = maximum (value : map max_tree_value subtrees)

sumSquares :: Integer -> Integer
sumSquares n = helper n 0
    where
    helper 0 acc = acc
    helper n acc = helper (n - 1) (acc + n ^ 2)